v1/plugins/logs/eventBuffer.go

```diff
@@ -0,0 +1,240 @@
+		return err
+	}
+	if int64(len(encoded)) > b.uploadSizeLimitBytes {
+		return fmt.Errorf("upload chunk size (%d) exceeds upload_size_limit_bytes (%d)",
+			int64(len(encoded)), b.uploadSizeLimitBytes)
+	}
+
+	return droppedNDCache{}
+}
+
+func push[T any](ch chan T, data T) {
+	select {
+	case ch <- data:
+	default:
+		<-ch
+		ch <- data
+	}
+}
+
+// Upload reads all the currently buffered events and uploads them as a gzip compressed JSON array to the service.
+// The events are uploaded in chunks limited by the uploadSizeLimitBytes.
+func (b *eventBuffer) Upload(ctx context.Context) (bool, error) {
+	b.payload.mtx.Lock()
+	defer b.payload.mtx.Unlock()
+
+	// This only blocks when the buffer is being reconfigured
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+
+	events := len(b.buffer)
+	for range events {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		case event := <-b.buffer:
+			if !b.payload.capacity(len(event), b.uploadSizeLimitBytes) {
+				ok, err := b.payload.upload(ctx, b.client, b.uploadPath)
+				if !ok || err != nil {
+					return false, err
+				}
+			}
+
+			if err := b.payload.add(event); err != nil {
+				return false, err
+			}
+		default:
+			break
+		}
+	}
+
+	return b.payload.upload(ctx, b.client, b.uploadPath)
+}
+
+// payloadBuffer contains the compressed JSON array to be uploaded
+type payloadBuffer struct {
+	buffer       *bytes.Buffer
+	mtx          sync.Mutex
+	writer       *gzip.Writer
+	closed       bool
+	bytesWritten int
+}
+
+func newPayloadBuffer() *payloadBuffer {
+	p := payloadBuffer{}
+	p.buffer = new(bytes.Buffer)
+	p.writer = gzip.NewWriter(p.buffer)
+	return &p
+}
+
+func (p *payloadBuffer) capacity(eventLen int, uploadSizeLimitBytes int64) bool {
+	return !(int64(eventLen+p.bytesWritten+1) > uploadSizeLimitBytes)
+}
+
+// add writes an event to the JSON array
+func (p *payloadBuffer) add(event []byte) error {
+	switch p.bytesWritten {
+	case 0: // Start new JSON array
+		n, err := p.writer.Write([]byte(`[`))
+		if err != nil {
+			return err
+		}
+		p.bytesWritten += n
+	default: // Append new event to JSON array
+		n, err := p.writer.Write([]byte(`,`))
+		if err != nil {
+			return err
+		}
+		p.bytesWritten += n
+	}
+
+	n, err := p.writer.Write(event)
+	if err != nil {
+		return err
+	}
+
+	p.bytesWritten += n
+
+	return nil
+}
+
+// upload closes the JSON array and attempts to send the data to the configured service
+func (p *payloadBuffer) upload(ctx context.Context, client rest.Client, uploadPath string) (bool, error) {
+	if p.bytesWritten == 0 {
+		return false, nil
+	}
+
+	if !p.closed {
+		_, err := p.writer.Write([]byte(`]`))
+		if err != nil {
+			return false, err
+		}
+
+		p.closed = true
+		if err := p.writer.Close(); err != nil {
+			return false, err
+		}
+	}
+
+	if err := uploadChunk(ctx, client, uploadPath, p.buffer.Bytes()); err != nil {
+		return false, err
+	}
+
+	p.bytesWritten = 0
+	p.buffer = new(bytes.Buffer)
+	p.writer = gzip.NewWriter(p.buffer)
+	p.closed = false
```

Comment:
[nit] Perhaps a p.reset() method would be nice? It could help convey the intention 🤔

Reply:
Sounds like a good idea to me, `reset` function added.